### PR TITLE
Add some lightweight typing

### PR DESCRIPTION
**Pull request by jvoisin**
This is a first quick pass over the codebase. If having better typing is something desirable, I'll do another more comprehensive one.

Having typing makes it easier to understand what's going on in the code, eg. "this function called `colorize_result` it taking either a `str` or `None`, and is returning either a `str` or `None`, so odds are that it's handling error conditions properly." It also makes the life of IDE/static analyzers/… easier.

### CODE REVIEW

**Comment by a13xp0p0v** [comment]
Hi @jvoisin, 
Thanks for the pull request! I like the idea.
There is a CI error, could you have a look, please?

File: kernel_hardening_checker/engine.py [hunk 1]
```diff
@@ -98,20 +98,20 @@ def check(self):
...
+    def json_dump(self, with_results: bool) ->dict:
         dump = {
             "option_name": self.name,
             "type": self.type,
             "desired_val": self.expected,
             "decision": self.decision,
             "reason": self.reason,
         }
-        if with_results:
+        if with_results and self.result:
```
Comment (a13xp0p0v) [inline kernel_hardening_checker/engine.py:114]: A question about this line.
All other changes add types in the function prototypes, but here you change the code.
Could you describe the reason?

**Comment by a13xp0p0v** [comment]
> Hi @jvoisin, Thanks for the pull request! I like the idea. There is a CI error, could you have a look, please?

I mean this error:
```
Traceback (most recent call last):
  File "bin/kernel-hardening-checker", line 13, in <module>
    import kernel_hardening_checker
  File "/home/runner/work/kernel-hardening-checker/kernel-hardening-checker/kernel_hardening_checker/__init__.py", line 32, in <module>
    def detect_arch(fname: str, archs: list[str]) -> tuple:
TypeError: 'type' object is not subscriptable
Error: Process completed with exit code 1.
```

**Comment by jvoisin** [comment]
The error is because I didn't realise Python ≤3.9 was supported

File: kernel_hardening_checker/engine.py (hunk 1 above)
Comment (jvoisin) [reply kernel_hardening_checker/engine.py]: > kernel_hardening_checker/engine.py:119: error: "None" has no attribute "startswith"  [attr-defined]

**Comment by a13xp0p0v** [comment]
@jvoisin, thanks!
 
The CI scripts of `kernel-hardening-checker` run on Python versions that are currently officially supported.
Should `kernel-hardening-checker` also support some older Python versions?

I've added a separate issue for this discussion: https://github.com/a13xp0p0v/kernel-hardening-checker/issues/133

File: kernel_hardening_checker/__init__.py [hunk 2]
```diff
@@ -24,13 +24,11 @@
...
-        f = gzip.open(file, *args, **kwargs)
-    else:
-        f = open(file, *args, **kwargs)
-    return f
+        return gzip.open(file, *args, **kwargs)
+    return open(file, *args, **kwargs)
 
 
-def detect_arch(fname: str, archs: List[str]) -> Tuple:
+def detect_arch(fname: str, archs: List[str]) -> Tuple[str | None, str]:
```
Comment (jvoisin) [inline kernel_hardening_checker/__init__.py]:
```suggestion
def detect_arch(fname: str, archs: List[str]) -> Tuple[Optional[str], str]:
```

**Comment by a13xp0p0v** [comment]
Hi @jvoisin,

Thank you for starting this work!

In this branch, I added the detailed static typing, which is checked with the `mypy` tool in CI.

Would you like to have a look and do a brief review?

I separated the commits that:
 - add typing,
 - fix mypy warnings,
 - do refactoring,
 - improve CI.
 
There should be no functional changes in this branch.
The `kernel-hardening-checker` output should be the same before these changes and after them.

**Comment by a13xp0p0v** [comment]
By the way, I just learned that `mypy` has the `--strict` mode and can generate the html report with coverage.

Do we need this, how do you think?

**Comment by jvoisin** [comment]
Wow, you really went all-in!

- I'm not sure it's worth duplicating the type checking step in several jobs. In fact, having it into its own job would make sense, so that it could be parallelized with the others.
- `-> None` is implicit, but I guess having it explicitly doesn't hurt.
- I don't think we really care about coverage: I added the typing annotations as always-up-to-date-comments, so that it's more clean what every function is doing/expecting. Reaching 100% typing information coverage will likely add way to many useless bloat.

Otherwise, LGTM :)

**Comment by a13xp0p0v** [comment]
> Wow, you really went all-in!
> 
> * I'm not sure it's worth duplicating the type checking step in several jobs. In fact, having it into its own job would make sense, so that it could be parallelized with the others.

Agree, fixed in https://github.com/a13xp0p0v/kernel-hardening-checker/pull/121/commits/dda21ff0d50bbe01acf7305946124e5d13d1bb3b

> * `-> None` is implicit, but I guess having it explicitly doesn't hurt.
> * I don't think we really care about coverage: I added the typing annotations as always-up-to-date-comments, so that it's more clean what every function is doing/expecting. Reaching 100% typing information coverage will likely add way to many useless bloat.
> 
> Otherwise, LGTM :)

Cool, thanks!

**Pull request merged**